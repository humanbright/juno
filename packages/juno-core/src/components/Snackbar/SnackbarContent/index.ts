export * from './SnackbarContent';
export * from './utils';
