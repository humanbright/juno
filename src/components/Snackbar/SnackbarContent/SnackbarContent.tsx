import MuiSnackbarContent from '@material-ui/core/SnackbarContent';
import React, {
  ComponentProps,
  forwardRef,
  isValidElement,
  useMemo,
} from 'react';

import {
  combineClasses,
  isRcElement,
  RcBaseProps,
  RcBaseSize,
  styled,
  withDeprecatedCheck,
  useThemeProps,
} from '../../../foundation';
import { RcCircularProgress } from '../../Progress/CircularProgress';
import { snackbarContentStyle } from './styles';
import {
  RcSnackbarContentClasses,
  RcSnackbarContentLoadingSizes,
} from './utils';

type RcSnackbarContentMessageAlignment = 'left' | 'center' | 'right';
type RcSnackbarContentSize = RcBaseSize<'small' | 'medium'>;
type RcSnackbarContentType = 'warn' | 'success' | 'error' | 'info';

type RcSnackbarContentProps = {
  /** message type */
  type?: RcSnackbarContentType;
  /** message size, default is `medium` */
  size?: RcSnackbarContentSize;
  /** message align  */
  messageAlign?: RcSnackbarContentMessageAlignment;
  /**
   * If `true`, the input will take up the full width of its container
   * when fullWidth the square will be `true`, if you still want square
   * set square={false}
   */
  fullWidth?: boolean;
  /** is use loading in action */
  loading?: boolean;
  /** @deprecated please use `aria-label` directly */
  ariaLabel?: string;
} & RcBaseProps<
  ComponentProps<typeof MuiSnackbarContent>,
  'variant' | 'elevation'
>;

/** @release */
const _RcSnackbarContent = forwardRef<any, RcSnackbarContentProps>(
  (inProps: RcSnackbarContentProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcSnackbarContent' });
    const {
      classes: classesProp,
      loading,
      size,
      messageAlign,
      type,
      fullWidth,
      action: actionProp,
      ...rest
    } = props;

    const classes = useMemo(
      () => combineClasses(RcSnackbarContentClasses, classesProp),
      [classesProp],
    );

    const action = useMemo(() => {
      const getItem = (item: any) => {
        if (isValidElement(item) && isRcElement(item, ['RcSnackbarAction'])) {
          /** when size is not equal default size, that mean user want to custom size */
          if (size !== RcSnackbarContent.defaultProps!.size) {
            return item;
          }

          return React.cloneElement(item as JSX.Element, { size });
        }
        return item;
      };

      if (actionProp instanceof Array) {
        return actionProp.map(getItem);
      }

      return getItem(actionProp);
    }, [actionProp, size]);

    return (
      <MuiSnackbarContent
        {...rest}
        ref={ref}
        classes={classes}
        action={
          loading ? (
            <RcCircularProgress
              color="inherit"
              size={RcSnackbarContentLoadingSizes[size!]}
            />
          ) : (
            action
          )
        }
      />
    );
  },
);

const RcSnackbarContent = styled(
  withDeprecatedCheck(
    _RcSnackbarContent,
    [
      {
        prop: 'ariaLabel',
        time: '2021-3',
        comment: `please use \`aria-label\` directly `,
      },
    ],
    'RcSnackbarContent',
  ),
)`
  ${snackbarContentStyle}
`;

RcSnackbarContent.defaultProps = {
  type: 'success',
  messageAlign: 'center',
  size: 'medium',
};

RcSnackbarContent.displayName = 'RcSnackbarContent';

export {
  RcSnackbarContent,
  RcSnackbarContentProps,
  RcSnackbarContentSize,
  RcSnackbarContentType,
};
