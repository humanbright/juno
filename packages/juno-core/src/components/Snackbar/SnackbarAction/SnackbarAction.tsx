import React, { ComponentProps, FunctionComponent, useMemo } from 'react';

import clsx from 'clsx';

import MuiButtonBase from '@material-ui/core/ButtonBase';

import {
  omit,
  RcBaseProps,
  RcClassesProps,
  RcPaletteProp,
  styled,
  useThemeProps,
} from '../../../foundation';
import { RcIcon, RcIconProps, SvgSymbol } from '../../Icon';
import { RcSnackbarContent, RcSnackbarContentSize } from '../SnackbarContent';
import { snackbarContentActionStyle } from './styles';
import { RcSnackbarActionClasses } from './utils';

type RcSnackbarActionProps = {
  /** variant of action */
  variant?: 'text' | 'icon';
  /** text color */
  color?: RcPaletteProp;
  /** size for action */
  size?: RcSnackbarContentSize;
  /** @deprecated should use symbol to replace that */
  icon?: SvgSymbol;
} & RcClassesProps<'icon' | 'text'> &
  RcBaseProps<ComponentProps<typeof MuiButtonBase>, 'color'> &
  Pick<RcIconProps, 'symbol'>;

const _RcSnackbarAction: FunctionComponent<RcSnackbarActionProps> = (
  inProps,
) => {
  const props = useThemeProps({ props: inProps, name: 'RcSnackbarAction' });
  const {
    children,
    variant,
    size,
    icon,
    symbol = icon,
    classes: classesProp,
    color,
    className: classNameProp,
    ...rest
  } = props;

  const className = useMemo(
    () =>
      clsx(classNameProp, {
        [RcSnackbarActionClasses.text]: variant === 'text',
        [RcSnackbarActionClasses.icon]: variant === 'icon',
      }),
    [classNameProp, variant],
  );

  const classes = useMemo(
    () => omit(classesProp, ['text', 'icon']),
    [classesProp],
  );

  return (
    <MuiButtonBase className={className} classes={classes} {...rest}>
      {variant === 'icon' ? (
        <RcIcon size={size} symbol={symbol}>
          {children}
        </RcIcon>
      ) : (
        children
      )}
    </MuiButtonBase>
  );
};

const RcSnackbarAction = styled(_RcSnackbarAction)`
  ${snackbarContentActionStyle}
`;

RcSnackbarAction.displayName = 'RcSnackbarAction';

RcSnackbarAction.defaultProps = {
  disableRipple: true,
  variant: 'text',
  size: RcSnackbarContent.defaultProps!.size,
};

export { RcSnackbarAction };
export type { RcSnackbarActionProps };
