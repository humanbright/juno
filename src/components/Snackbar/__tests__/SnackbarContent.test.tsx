import React from 'react';

import { shallowWithTheme } from '../../../../tests';
import { RcSnackbarAction } from '../SnackbarAction/SnackbarAction';
import { RcSnackbarContent, RcSnackbarContentProps } from '../SnackbarContent';

describe('SnackbarContent', () => {
  let defaultProps: RcSnackbarContentProps;
  beforeEach(() => {
    jest.clearAllMocks();
    defaultProps = {
      message: 'default text',
    };
  });
  it('should render', () => {
    const wrapper = shallowWithTheme(<RcSnackbarContent {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('Align action buttons to the first row when toast content with action and multiple lines of text [JPT-4487]', () => {
    const actions = [<RcSnackbarAction key="1">test</RcSnackbarAction>];
    const wrapper = shallowWithTheme(
      <RcSnackbarContent {...defaultProps} action={actions} />,
    );
    expect(
      wrapper
        .find(RcSnackbarContent)
        .at(0)
        .props().action,
    ).not.toEqual(null);
    expect(wrapper).toMatchSnapshot();
  });
});
