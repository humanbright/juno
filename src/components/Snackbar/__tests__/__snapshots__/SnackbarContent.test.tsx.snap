// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarContent Align action buttons to the first row when toast content with action and multiple lines of text [JPT-4487] 1`] = `
<StylesProvider
  injectFirst={true}
>
  <Component>
    <RcSnackbarContent
      action={
        Array [
          <RcSnackbarAction
            disableRipple={true}
            size="medium"
            variant="text"
          >
            test
          </RcSnackbarAction>,
        ]
      }
      message="default text"
      messageAlign="center"
      size="medium"
      type="success"
    />
  </Component>
</StylesProvider>
`;

exports[`SnackbarContent should render 1`] = `
<StylesProvider
  injectFirst={true}
>
  <Component>
    <RcSnackbarContent
      message="default text"
      messageAlign="center"
      size="medium"
      type="success"
    />
  </Component>
</StylesProvider>
`;
